build/gen_stub.php

```diff
@@ -1976,14 +2031,18 @@ public function discardInfoForOldPhpVersions(): void {
      */
     public function getDeclaration(iterable $allConstInfos): string
     {
-        $type = $this->phpDocType->tryToSimpleType();
-        if ($type === null || !$type->isBuiltin) {
-            throw new Exception("Constant " . $this->name->__toString() . " must have a simple built-in type");
+        $simpleType = ($this->phpDocType ?? $this->type)->tryToSimpleType();
+        if ($simpleType && $simpleType->name === "mixed") {
+            $simpleType = null;
+        }
+
+        $value = EvaluatedValue::createFromExpression($this->value, $simpleType, $this->cValue, $allConstInfos);
+        if ($value->isUnknownConstValue && ($simpleType === null || !$simpleType->isBuiltin)) {
+            throw new Exception("Constant " . $this->name->__toString() . " must have a built-in PHPDoc type as the type couldn't be inferred from its value");
         }
 
-        $value = EvaluatedValue::createFromExpression($this->value, $type, $this->cValue, $allConstInfos);
         // i.e. const NAME = UNKNOWN;, without the annotation
-        if ($value->isUnknownConstValue && $this->cValue === null && $value->expr instanceof Expr\ConstFetch && $value->expr->name->__toString() == $this->name->__toString()) {
```

Comment:
`$value->expr->name->__toString() == $this->name->__toString()` was a bug in the original implementation:
so that the exception below was never thrown. I'm not sure though why it's needed in the condition, but changed it to the correct one in any case.